File: serialport/serialport.cpp
#include "serialport.h"

QStringList SerialPort::baudList;
Init init;

SerialPort::SerialPort(QObject *parent):QObject(parent)
{
    serialPort = new QSerialPort(this);
}

SerialPort::~SerialPort()
{
}

/**
 * @brief 获取pc中可连接的com
 *
 * @return QStringList
 */
QStringList SerialPort::getSerialPortList()
{
    QStringList strList;
    foreach(const QSerialPortInfo &SerialPortInfo,QSerialPortInfo::availablePorts())
    {
        strList.append(SerialPortInfo.portName());
    }
    return strList;
}


/**
 * @brief 简化版槽函数
 *
 * @param reciever
 * @param slotRec
 */
void SerialPort::connectRec(QObject *reciever, const char *slotRec)
{
    this->connect(serialPort,SIGNAL(readyRead()),reciever,slotRec);
}


/**
 * @brief 打开指定端口的串口
 *
 * @param port
 * @param baudRate
 * @return bool
 */
bool SerialPort::openPort(const QString &port, int baudRate)
{
    bool isOpenSuccess;
    serialPort->setPortName(port);
    isOpenSuccess = (serialPort->open(QIODevice ::ReadWrite)
                     //串口参数设置
                     &&serialPort->setBaudRate(baudRate)
                     &&serialPort->setDataBits(QSerialPort::Data8)
                     &&serialPort->setDataErrorPolicy(QSerialPort::IgnorePolicy)
                     &&serialPort->setFlowControl(QSerialPort::NoFlowControl)
                     &&serialPort->setParity(QSerialPort::NoParity)
                     &&serialPort->setStopBits(QSerialPort::OneStop));
    return isOpenSuccess;
}

/**
 * @brief 发送函数及其各种重载
 *
 * @param dataStr
 */
void SerialPort::send(const QString &dataStr)
{
    serialPort->write(dataStr.toLatin1());
}

void SerialPort::send(QByteArray &data)
{
    serialPort->write(data);
}

/**
 * @brief 接收数据, 须配合connectRec
 *
 * @return QByteArray
 */
QByteArray SerialPort::read()
{
    return serialPort->readAll();
}

/**
 * @brief 获取当前连接的串口名
 *
 * @return QString
 */
QString SerialPort::getCurPortName()
{
    return serialPort->portName();
}

void SerialPort::close()
{
    serialPort->close();
}

File: SerialPort.py
from PyQt5.QtCore import QObject, QIODevice, QByteArray
from PyQt5.QtSerialPort import QSerialPort, QSerialPortInfo

baudList = ('1200', '4800', '9600', '14400', '19200', '38400', '57600', '115200')


class SerialPort(QObject):
    def __init__(self):
        super(SerialPort, self).__init__()
        self.serialPort = QSerialPort(self)

    @staticmethod
    def getSerialPortList():
        """
        当前所连接的串口列表
        :return: list-of-QSerialPortInfo
        """
        portlist = list(map(lambda port: port.portName(), QSerialPortInfo.availablePorts()))
        return portlist

    def connectRec(self, slotRec):
        self.serialPort.readyRead.connect(slotRec)

    def openPort(self, portName, baudRate):
        self.serialPort.setPortName(portName)
        isOpenSuccess = (self.serialPort.open(QIODevice.ReadWrite)
                         and self.serialPort.setBaudRate(baudRate)
                         and self.serialPort.setDataBits(QSerialPort.Data8)
                         and self.serialPort.setDataErrorPolicy(QSerialPort.IgnorePolicy)
                         and self.serialPort.setFlowControl(QSerialPort.NoFlowControl)
                         and self.serialPort.setParity(QSerialPort.NoParity)
                         and self.serialPort.setStopBits(QSerialPort.OneStop))
        return isOpenSuccess

    def send(self, data):
        if len(data) == 0 or data is None:
            return
        if isinstance(data, str):
            return self.serialPort.write(QByteArray(bytes(data)))
        else:
            return self.serialPort.write(QByteArray(data))

    def read(self):
        """
        读取串口收到的数据
        :return: bytes
        """
        return self.serialPort.readAll().data()

    def getCurPortName(self):
        return self.serialPort.portName()

    def close(self):
        self.serialPort.close()
